internal/app/backend/backend_service.go

```diff
@@ -95,59 +101,53 @@ func (s *backendService) FetchMatches(req *pb.FetchMatchesRequest, stream pb.Bac
 	}
 
 	ctx := stream.Context()
-	matchChan := make(chan *pb.Match)
-	errChan := make(chan error)
-
-	go func(matchChan chan<- *pb.Match, errChan chan<- error) {
-		var wg sync.WaitGroup
-
-		defer func() {
-			wg.Wait()
-			close(matchChan)
-			close(errChan)
-		}()
-
-		for _, profile := range req.Profile {
-			wg.Add(1)
-			go func(profile *pb.MatchProfile) {
-				defer wg.Done()
+	resultChan := make(chan mmfResult, len(req.Profile))
+
+	for _, profile := range req.Profile {
+		go func(profile *pb.MatchProfile) {
+			var matches []*pb.Match
+			// Get the match results that will be sent.
+			// TODO: The matches returned by the MatchFunction will be sent to the
+			// Evaluator to select results. Until the evaluator is implemented,
+			// we channel all matches as accepted results.
+			switch (req.Config.Type).(type) {
+			case *pb.FunctionConfig_Grpc:
+				matches, err = matchesFromGRPCMMF(ctx, profile, grpcClient)
+			case *pb.FunctionConfig_Rest:
+				matches, err = matchesFromHTTPMMF(ctx, profile, httpClient, baseURL)
+			}
+			resultChan <- mmfResult{matches, err}
+		}(profile)
+	}
 
-				// Get the channel over which the generated match results will be sent.
-				switch (req.Config.Type).(type) {
-				case *pb.FunctionConfig_Grpc:
-					matchesFromGRPCMMF(ctx, profile, grpcClient, matchChan, errChan)
-				case *pb.FunctionConfig_Rest:
-					matchesFromHTTPMMF(ctx, profile, httpClient, baseURL, matchChan, errChan)
-				}
-			}(profile)
-		}
-	}(matchChan, errChan)
+	proposals := []*pb.Match{}
 
-	for {
+	for i := 0; i < len(req.Profile); i++ {
 		select {
 		case <-ctx.Done():
```

Comment:
returning early here will deadlock goroutines which are trying to send results on resultChan.